packages/tools/guiEditor/src/workbenchEditor.tsx

```diff
@@ -264,39 +333,108 @@ export class WorkbenchEditor extends React.Component<IGraphEditorProps, IGraphEd
+                            </div>
+                        }
+                        {
+                            <div className={"toolbar-content-sub1"}>
+                                {ControlTypesFHalf.map((type) => {
+                                    return (
+                                        <div
+                                            className={"toolbar-label"}
+                                            key={type.className}
+                                            onDragStart={() => {
+                                                this._draggedItem = type.className;
+                                            }}
+                                            onClick={() => {
+                                                this.onCreate(type.className);
+                                            }}
+                                            title={type.className}
+                                        >
+                                            {type.icon && (
+                                                <div className="toolbar-icon" draggable={true}>
+                                                    <img src={type.icon} alt={type.className} width="40px" height={"40px"} />
+                                                </div>
+                                            )}
+                                        </div>
+                                    );
+                                })}
+                                {ControlTypesSHalf.map((type) => {
+                                    return (
+                                        <div
+                                            className={"toolbar-label"}
+                                            key={type.className}
+                                            onDragStart={() => {
+                                                this._draggedItem = type.className;
+                                            }}
+                                            onClick={() => {
+                                                this.onCreate(type.className);
+                                            }}
+                                            title={type.className}
+                                        >
+                                            {type.icon && (
+                                                <div className="toolbar-icon" draggable={true}>
+                                                    <img src={type.icon} alt={type.className} width="40px" height={"40px"} />
+                                                </div>
+                                            )}
+                                        </div>
+                                    );
+                                })}
+                                {}
+                            </div>
+                        }
+                    </div>
+                </>
+            );
+        }

```

Comment:
It seems like this code repeats itself quite a lot. You can set the different properties before returning the JSX, and  then return a single implementation of the same JSX. for example, something like this:

```javascript
// ...
const controlTypeMap = this.state.toolbarExpand ? ControlTypes : [...ControlTypesFHalf, ...ControlTypesSHalf];
const idForElement = this.state.toolbarExpand ? "toolbarGrab" : "toolbarGrabExpanded";
``` 

and then use those variables to render the toolbar.